C++ Support

misson20000 opened the pull request:
Closes #17. Credit to @dvdfreitag

misson20000 committed:
implement a few more syscalls

dvdfreitag committed:
Initial C++ support

These libraries compile fine, but test_cpp fails with linker errors. lld is complaining about some missing symbols that really aren't missing, like main, and some pthread functions. AFAICT the catalog functions (catgets etc) are not defined anywhere yet.

lubosz committed:
phtread.h: use extern C.

lubosz committed:
libtransitor.mk: Set -error-limit=0 on LD_FLAGS.

lubosz committed:
include/stdlib.h: use extern C.

roblabla committed:
Use the stdlib.h from newlib, include the configured newlib.h

roblabla committed:
Handle the exception handling in linker script

roblabla committed:
Ugly hack to run the tests

roblabla committed:
Enhanced debugging in elf2nxo.

roblabla committed:
Better debug info.

dvdfreitag committed:
Merge pull request #2 from roblabla/feature-betterCPP

Feature better cpp

DavidBuchanan314 committed:
Fix relocations, implement R_AARCH64_ABS64 relocation

dvdfreitag committed:
Merge pull request #3 from DavidBuchanan314/master

Fix relocations, implement R_AARCH64_ABS64 relocation

dvdfreitag committed:
Add CXXFLAGS

For compatability reasons, CXXFLAGS has been added alongside CXX_FLAGS

dvdfreitag committed:
Fix C++ Exceptions

The reason that exceptions were causing runtime crashes is because they weren't aligned to 4k in the linker script.

misson20000 committed:
Merge branch 'development' into feature-cpp

misson20000 committed:
Merge branch 'development' into feature-cpp

misson20000 committed:
put .gcc_except_table in .rodata to work around nxodance being fragile

misson20000 commented on link.T:
```diff
@@ -63,20 +63,18 @@ NORELOC_BSS_END_ = .;
+    *(.eh_frame_hdr)
+    __eh_frame_hdr_end = .;
+  }
+
```
These sections still don't make it into the final NRO. They get discarded by elf2nxo.py

dvdfreitag replied:
Interesting, so the runtime errors we were seeing previously were simply an alignment issue, probably throwing _everything_ out of wack.

I assume this is some issue with elf2nxo, or nxodance?

dvdfreitag commented:
Something in the merge with development broke the libcxx build, specifically it looks like an include path for `strtold` got broken.

misson20000 committed:
fix cpp build

dvdfreitag committed:
Ignore xz Automake cache

Automake creates a directory called autom4te which contains cached build outputs. It should not be checked in.

dvdfreitag committed:
Fix openlibm build

Remove unneeded files and fix openlibm.mk so that the correct source files are found, build, and linked into build/openlibm/libm.a

dvdfreitag committed:
Delete openlibm.h

openlibm.h is unnecessary for our purposes, it's a holdover from the old include system. It exists so that the include paths aren't poisoned with extra `math.h` or `complex.h`

roblabla committed:
Merge pull request #99 from dvdfreitag/feature-cpp

Fix openlibm build

roblabla commented:
Related issue: https://github.com/reswitched/libtransistor/pull/98

roblabla commented:
Investigation done by dvdfreitag:

```
dvdfreitag: So I think I know why exceptions aren't working
dvdfreitag: 

int square(int num)
{
    try
    {
        throw num;
    }
    catch (int a)
    {
        return a;
    }
}

dvdfreitag: That should produce the following output:
dvdfreitag: 

_Z6squarei:                             // @_Z6squarei
        str     x19, [sp, #-32]!        // 8-byte Folded Spill
        stp     x29, x30, [sp, #16]     // 8-byte Folded Spill
        add     x29, sp, #16            // =16
        mov      w19, w0
        orr     w0, wzr, #0x4
        bl      __cxa_allocate_exception
        str       w19, [x0]
        adrp    x1, :got:_ZTIi
        mov      x2, xzr
        ldr     x1, [x1, :got_lo12:_ZTIi]
        bl      __cxa_throw
        bl      __cxa_begin_catch
        ldr       w19, [x0]
        bl      __cxa_end_catch
        ldp     x29, x30, [sp, #16]     // 8-byte Folded Reload
        mov      w0, w19
        ldr     x19, [sp], #32          // 8-byte Folded Reload
        ret

dvdfreitag: But, in libtransistor all of __cxa_allocate_exception etc are replaced with #0
dvdfreitag: 

_Z6squarei:
    str    x19, [sp, #-32]!
    stp    x29, x30, [sp, #16]
    add    x29, sp, #16
    mov     w19, w0
    orr    w0, wzr, #0x4
    bl    #0
    str        w19, [x0]
    adrp    x1, #0
    mov     x2, xzr
    ldr        x1, [x1]
    bl    #0
    bl    #0
    ldr        w19, [x0]
    bl    #0
    ldp    x29, x30, [sp, #16]
    mov     w0, w19
    ldr    x19, [sp], #32
    ret

(edited)
dvdfreitag: Looks like libcxxabi.a doesn't contain any of those symbols :thinking:
dvdfreitag: Ah wait, I was wrong. They're there
dvdfreitag: mmm, interesting. When you link to an nro they are changed from zero
misson20000: (probably our link script being lame)
dvdfreitag: So __cxa_allocate_exception is replaced with #346276
dvdfreitag: Which is way out beyond the end of the nro, 5ba54 is the last address in text
dvdfreitag: Oh, actually
dvdfreitag: :face_palm:
dvdfreitag: So that's jumping to 548a4, which is somewhere in the middle of snprintf
dvdfreitag: __cxa_allocate_exception is in the nro, but at a totally different location
dvdfreitag: 

__cxa_allocate_exception:
   54960:

dvdfreitag: An offset of -188
dvdfreitag: Yeah it's not a fixed offset, __cxa_throw is -208 away from where it's actually at
dvdfreitag: And the type information symbols aren't present, eg _ZTIi which is out at 0x75000
dvdfreitag: Could this possibly be due to an issue with the modifications made to nxodance?
misson20000: It sounds like an nxodance issue, yeah
dvdfreitag: :/
dvdfreitag: 

Function                    Address        Real Address    Offset
__cxa_allocate_exception    0x548A4        0x54960         -0xBC
__cxa_throw                 0x5493C        0x54A0C         -0xD0
__cxa_begin_catch           0x54A24        0x54AF8         -0xD4
__cxa_end_catch             0x54AB8        0x54b94         -0xDC


dvdfreitag: Retr0id roblabla Looks like the issue with exceptions is something to do with the modifications that were made to nxodance ^
```

misson20000 committed:
Merge branch 'development' into feature-cpp

misson20000 committed:
move module header offset at NRO+0x4 to crt0 to workaround cmake toolchain tests failing

misson20000 committed:
Merge branch 'development' into feature-cpp

misson20000 committed:
pass --eh-frame-hdr to linker and stop swallowing sections into the wrong place (.eh_frame_hdr was getting pulled into .eh_frame because it was matched by .eh_frame*)

misson20000 committed:
add tests for libunwind, cpp exceptions, and cpp dynamic memory

misson20000 committed:
Merge branch 'development' into feature-cpp
